### ISSUE TITLE
(Stack Overflow) is there a recommended pattern for migrating fields?

### ISSUE THREAD

**Issue by lukepighetti** [google/built_value.dart#1118]
Let's say you have a built value with a field with wireName `src`, and you need to change that wire name to `source`. What is the recommended method for migrating this field such that the class can still be serialized with `src` and `source` fields, but only outputs `source` field?

**Comment by davidmorgan** [comment]
You could create an optional field with wirename `src`, then use a builder hook 

https://pub.dev/documentation/built_value/latest/built_value/BuiltValueHook-class.html

to copy this field into the `source` field if it's present, and clear it, something like

```
BuiltValueHook(finalizeBuilder: true)
static void initialize(FooBuilder b) {
  if (b.src != null) {
    b.source = b.src;
    b.src = null;
  }
}
```

The other way would be to write a custom serializer or serializer plugin, giving you complete control over what's serialized.

**Comment by lukepighetti** [comment]
Is there room for a PR to add a more opinionated migration pattern to built_value? Any idea what that might look like? I could put together a proposal, but guard rails would be appreciated

**Comment by davidmorgan** [comment]
Hmmmmm we could add an `oldWireNames` option to `@BuiltValueField` where you can specify the old names? Then the serializer would recognize the old names but never write them. That sounds like a nice feature.

**Comment by lukepighetti** [comment]
Yeah, that's almost exactly what I was thinking. `@BuiltValueField(wireName: 'source', oldWireNames: ['src'])`

**Comment by davidmorgan** [comment]
I think the change to the generated code will be relatively straightfoward; the only awkward point I see is what to do if both 'source' and one or more old names appear in the JSON. For that I suggest keeping it simple: the new 'wireName' always wins, there is no guaranteed winner between multiple 'oldWireNames', and never check/throw.

You're welcome to take a shot at it :)

The main thing to know when working on built_value_generator is you'll need to run tool/use_local_deps to switch to using local versions of all the packages, so you can use/add end to end coverage.

Please note I'll be on vacation and offline on Friday and next week.

Thanks!
